metals/src/main/scala/scala/meta/internal/metals/MultilineStringFormattingProvider.scala

```diff
@@ -95,30 +112,61 @@ final class OnTypeFormattingProvider(
     shouldAddPipes
   }
 
-  def format(
-      params: DocumentOnTypeFormattingParams
-  ): Future[java.util.List[TextEdit]] = {
-    val source = params.getTextDocument.getUri.toAbsolutePath
-    val range = new Range(params.getPosition, params.getPosition)
-
-    val edit = if (source.exists) {
+  private def withToken(
+      textId: TextDocumentIdentifier,
+      range: Range,
+      newlineAdded: Boolean
+  )(
+      fn: (String, meta.Position) => List[TextEdit]
+  ): Future[List[TextEdit]] = Future {
+    val source = textId.getUri.toAbsolutePath
+    if (source.exists) {
       val sourceText = buffer.get(source).getOrElse("")
-      val pos = params.getPosition.toMeta(
+      val pos = range.getStart.toMeta(
         Input.VirtualFile(source.toString(), sourceText)
       )
-      if (pipeInScope(pos, sourceText)) {
+      if (pipeInScope(pos, sourceText, newlineAdded)) {
         val tokens =
           Input.VirtualFile(source.toString(), sourceText).tokenize.toOption
-        tokens.flatMap { tokens: Tokens =>
-          if (multilineStringInTokens(tokens, pos, sourceText)) {
-            Some(new TextEdit(range, indent(sourceText, pos) + "|"))
-          } else {
-            None
-          }
+        tokens.toList.flatMap { tokens: Tokens =>
+          if (multilineStringInTokens(tokens, pos, sourceText))
+            fn(sourceText, pos)
+          else Nil
         }
-      } else None
-    } else None
-    Future.successful(edit.toList.asJava)
+      } else Nil
+    } else Nil
   }
 
+  def format(
+      params: DocumentOnTypeFormattingParams
+  ): Future[List[TextEdit]] = {
+    val range = new Range(params.getPosition, params.getPosition)
+    withToken(
```

Comment:
formatting looks a bit stretched. can something be done with it? e.g. extracting parameters to variables